### DISCUSSION TITLE
Check For Websocket Data

### DISCUSSION THREAD

**Discussion by limitedAtonement** [miguelgrinberg/flask-sock#97] [Q&A]
The handler in `examples/clock.py` looks like this:

```python
@sock.route('/clock')
def clock(ws):
    client_list.append(ws)
    while True:
        data = ws.receive()
        if data == 'stop':
            break
    client_list.remove(ws)
```

When terminating the server, the termination is held up because of the blocking call on `ws.receive()`. It would be better if we could do something like

```python
@sock.route('/clock')
def clock(ws):
    client_list.append(ws)
    while keep_running():
        if ws.available():
            data = ws.receive()
            if data == 'stop':
                break
        else:
            time.sleep(.5)
    client_list.remove(ws)
```

`keep_running` can be set up to return `False` if a `SIG_TERM` was sent or whatever.

Is there any way to get this type of non-blocking behavior?

**Comment by miguelgrinberg** [comment]
What happens when SIGTERM is received varies depending on the web server that you are using. A proper handling of SIGTERM by the web server would close all open handles, which would in turn raise an exception on the `receive()` call. Not sure what web server you are using, but I would prefer to not add code in this package to compensate for web servers that don't do this correctly.

**Comment by limitedAtonement** [comment]
A just response. I'm using gunicorn (without gevent and eventlet at the moment). I'll see what it takes to get gunicorn to close IO handles on SIGTERM.
